docs/proposals/workflow.md

```diff
@@ -0,0 +1,124 @@
+
+## Implementation
+
+The basic idea is to add a label selector to the current Job API object. The new selector will determine the parent jobs.  The parent jobs are the jobs current job will depend on. The current job will be scheduled once all the parent jobs will run to completion.
+The strongest point in favor of this approach is the ability to re-use the current `job` implementation. Using a label selector the current `job` will simply become a `graph` of jobs: all the features/controllers that will be built on top of `job` like for example [#11980](https://github.com/kubernetes/kubernetes/issues/11980) and [#17242](https://github.com/kubernetes/kubernetes/issues/17242) will be automatically extended to support workflows. Not sure about [#16845](https://github.com/kubernetes/kubernetes/issues/16845) and [#14188](https://github.com/kubernetes/kubernetes/issues/14188).
+A similar approach is implemented in Chronos for [dependent jobs](https://mesos.github.io/chronos/docs/api.html#adding-a-dependent-job).
+
+### API Object
+
+To implement _workflows_ the `Job` API should be modified:
+
+```go
+// JobSpec describes how the job execution will look like.
+
+// These are valid conditions of a job.
+const (
+	// JobComplete means the job has completed its execution.
+	JobComplete JobConditionType = "Complete"
+	// JobWaiting means the job is waiting for its parents to finish their tasks.
+	JobWaiting  JobConditionType = "Waiting"
+)
+
+type JobSpec struct {
+	...
+	// Job labels selector to detect parent jobs.
+	ParentSelector map[string]string `json:"parentSelector"`
+}
+
+```
+
+#### JobSpec
+
+A new labels selector will be added to `job.spec`. The `job.Spec.ParentSelector` is a label query over a set of jobs.
+If all selected jobs are completed the `job` will be started immediately.
+If `job.Spec.ParentSelector` is absent the `job` will be started immediately.
+If `job.Spec.ParentSelector` is pointing to a non-existing job, the job will wait indefinitely.
+Since labels are forwarded directly from `job` to pods, users should not create pods whose labels match this selector, either directly,
+via another Job, or via another controller (for example Replication controller), [see](https://github.com/kubernetes/kubernetes/issues/14961).
+
+#### JobStatus
+
+`job.Status` won't be modified, but a new job condition will be added.
+
+#### JobCondition
+
+A new constant value will be added to `JobConditionType` - `Waiting`. This will inform the job is waiting for its parents to finish execution. `Waiting` condition will be valid only for `workflow job`.
+
+### CRUD
+
+* Creating a job without a `job.Spec.Parentselector` (nil `job.Spec.Parentselector`) will simply follow the usual `job` life cycle.
+* Without a `job.Spec.ParentSelector` (i.e. nil) the job will be started immediately.
+* With a non nil `job.Spec.ParentSelector` the logic already described will be considered.
+* The validation will prevent the user to set an _empty_ `job.Spec.ParentSelector`.
```

Comment:
The validation will prevent the user from setting an _empty_ `job.spec.parentSelector` and pointing it to the same job.